plugins/modules/api.py

```diff
@@ -454,11 +455,11 @@ def api_arbitrary(self):
             for i in arbitrary_result:
                 self.result['message'].append(i)
             self.return_result(False)
-        except Exception as e:
+        except LibRouterosError as e:
             self.errors(e)
 
     def return_result(self, ch_status=False, status=True):
```

Comment:
From def errors we return True when we get TrapError, I do that with purpose. No easy way to do pre check via api for different tasks in ROS ! For this reason I ask for example in #32. 

for this approach (original code)
**Pros:** No need to handle **false positive** failers (we have such in ros)
**Cons:** Yes we will get fail as is excepted ! however for the  **false positive**  cases users must handle them via roles/playbooks 

For example: 
`    "msg": "['failure: already have such address']"
`
This happen when you try to add same ip address to interface .. first run will be successful, second run will fail !
Now I know this is my strange opinion, however this save overhand for such false positive cases like add ip address.

If we allow to FAIL in such case, the role / playbook must have SPECIFIC task to check if such address already exist end  etc .. However this is only for add ip what about all other basilion different tasks in ros. For each different task playook/role must have specific pre-check 

I hope you get my point way I do that in first place .. Over all I'm ok to make it fail (change must be done in  def errors() ) Probably additional module var to Fail or Not to Fail will be a good idea :)

    def errors(self, e):
        if e.__class__.__name__ == 'TrapError':
            self.result['message'].append("%s" % e)
            self.return_result(False, **True**)
        self.result['message'].append("%s" % e)
        self.return_result(False, False)

**NOTE: I'm really sorry for this late review :(**

Reply:
When I think more, probably my opinion is wrong and we should make it to fail.

Reply:
one more think: 

Is the correct approach to make it "Fail" -- I fully agree however this will make idempotent of the roles nightmare, the reason is because ros don't make a difference between foe example "wrong ip address mask" and "ip address already exist " (as example).
For this reason this module don't support check_mode 

May be I miss something and more research is needed ... 
I really don't know ...

Reply:
```
[admin@chr7b] > ip address/add address=192.168.12.3 netmask=255.255.256.0 interface=ether2
invalid value for argument netmask
[admin@chr7b] > ip address/add address=192.168.12.3 netmask=255.255.255.0 interface=ether2  
[admin@chr7b] > ip address/add address=192.168.12.3 netmask=255.255.255.0 interface=ether2 
failure: already have such address
```


same is via api module if we make it to fail
**first run:** ansible-playbook playbooks/ros_ip_addr_add.yml --limit chr7b --ask-vault-pass -vvvv

```
ok: [chr7b] => {
    "msg": {
        "changed": true,
        "failed": false,
        "msg": [
            "added: .id= *E"
        ]
    }
}
```

**second run:** ansible-playbook playbooks/ros_ip_addr_add.yml --limit chr7b --ask-vault-pass -vvvv

```
    "changed": false,
    "invocation": {
        "module_args": {
            "add": "broadcast= comment= copy-from= disabled=no netmask=255.255.255.0 network=192.168.192.0 address=192.168.192.20 interface=ether2",
            "ca_path": null,
            "cmd": null,
            "hostname": "chr7b",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "path": "ip address",
            "port": null,
            "query": null,
            "remove": null,
            "ssl": "false",
            "tls": false,
            "update": null,
            "username": "admin",
            "validate_cert_hostname": false,
            "validate_certs": true
        }
    },
    "msg": "['failure: already have such address']"
}

```

**wrong mask**  ansible-playbook playbooks/ros_ip_addr_add.yml --limit chr7b --ask-vault-pass -vvvv -e "ros_ip_addr_netmask=255.255.25.256"

```
    "changed": false,
    "invocation": {
        "module_args": {
            "add": "broadcast= comment= copy-from= disabled=no netmask=255.255.25.256 network=192.168.192.0 address=192.168.192.20 interface=ether2",
            "ca_path": null,
            "cmd": null,
            "hostname": "chr7b",
            "password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "path": "ip address",
            "port": null,
            "query": null,
            "remove": null,
            "ssl": "false",
            "tls": false,
            "update": null,
            "username": "admin",
            "validate_cert_hostname": false,
            "validate_certs": true
        }
    },
    "msg": "['invalid value for argument netmask']"
}

```

Reply:
> Is the correct approach to make it "Fail" -- I fully agree however this will make idempotent of the roles nightmare, the reason is because ros don't make a difference between for example "wrong ip address mask" and "ip address already exist " .
> For this reason this module don't support check_mode
> 
> May be I miss something and more research is needed ...
> I really don't know ...

Well, that's always the problem with low-level modules like `ansible.builtin.command` and `community.routeros.command` - it's a non-trivial amount of work to make them do something idempotent. If a command executed on a host (for `ansible.builtin.command`) or a router (for `community.routeros.command`) fails, the module should fail, and the user should be able to decide how to react on the failure. Simplify making the module ignore errors by default is definitely not right, since the module has no idea whether an error was intended, acceptable, or not.

Ansible provides `failed_when` for tasks (https://docs.ansible.com/ansible/latest/user_guide/playbooks_error_handling.html#defining-failure) to allow role/playbook authors to define when a command really failed. This can be used to distinguish between "acceptable errors" and "unacceptable errors". In your above example, `failed_when` can be something like `result is success or (result is failed and 'failure: already have such address' in result.msg)`. That way, you can make the module invocation idempotent, and still report errors on wrong input.

In the long run, it would be great if there would be more advanced modules for such tasks, which know about what they run and how to handle it - see #33.

Reply:
I agree in future will check with librouteros if is possible / usefully to utilize https://wiki.mikrotik.com/wiki/Manual:API#category somehow

Reply:
I propose to explicitly mention the case of "already have ..." errors that will now be reported in the changelog, since that could be considered a regression.

Reply:
The changelog fragment definitely needs to be extended. I'm wondering whether we should make this even more prominent, or whether this should even be a change for community.routeros 2.0.0.